### ISSUE TITLE
Yolov8-OBB Distorted Validation Labels

### ISSUE THREAD

**Issue by CedricLienhard** [ultralytics/ultralytics#9555]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Val

### Bug

Hello, 

I've playing around with the yolov8-obb model and I have to say that it works pretty well on the given labelled data (dota dataset for example). Congrats for releasing a model that is finally able to achieve this in a reasonable processing time!

However, I've then tried to make it work on my own dataset and the strange thing that I experiment is that the **validation** label are distorted, but **not the training** labels. Please see below example, where I intentionally put a huge bounding box around the ship and then put the same image in both sets (train and val) and here is what is shown: 

train image with label drawn on it => it is a rectangle, as expected
![train_batch1](https://github.com/ultralytics/ultralytics/assets/49435825/55d137f6-3410-490d-8a33-0174d0fd1f78)

validation image with label drawn on it => this one has the exact same label as the train image, but we get a slightly "skewed" rectangle!  The effect is much more pronounced on other images I have but I can't show them for confidentiality reasons.
![val_batch1_labels](https://github.com/ultralytics/ultralytics/assets/49435825/d7e491ad-deec-40b9-952e-58d30df66ef1)

My question is the following: is yolov8_obb using these distorted labels for validation (which would be very wrong) or is this just a display artifact that arises for whatever reason? 

The below reported issue also explains the problem but no solution was given.

https://discuss.roboflow.com/t/yolov8-obb-generated-labels-are-not-rectangles-and-do-not-match-the-original-oriented-rectangles/4770

### Environment

Ultralytics YOLOv8.1.24 🚀 Python-3.9.18 torch-2.2.1+cu118 CUDA:0 (NVIDIA GeForce RTX 3070 Laptop GPU, 7966MiB)
Setup complete ✅ (16 CPUs, 62.6 GB RAM, 375.4/431.5 GB disk)

### Minimal Reproducible Example

model = YOLO('yolov8n-obb.pt')  # load an official model
results = model.train(data=data_path, epochs=10, imgsz=640, device=0, batch=1, cfg = hyp_file)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@CedricLienhard hello there! 👋

Thanks for sharing your experience and for the kind words about our YOLOv8-OBB model! It's great to hear you've had success with it on the Dota dataset.

Regarding the issue with distorted validation labels, it sounds like what you're seeing could indeed be a display artifact rather than an issue with how the model is using the labels for validation. This might stem from the visualization step rather than the validation itself. 

To better diagnose and possibly rectify this, could you try adding `--save-conf` and `--save-txt` flags during your validation to save the outputs and compare the predicted bounding boxes against your original annotations? This can be particularly helpful to understand if the issue persists in the saved outputs or if it's just during the visualization.

```bash
model.val(data=data_path, imgsz=640, half=False, save_conf=True, save_txt=True)
```

If the saved outputs match your expectations but the visualization doesn't, we're likely dealing with a visualization issue. In that case, explore the visualization code or library you're using to draw the rectangles, ensuring it handles oriented bounding boxes (OBB) correctly. 

Again, thanks for bringing this up, and we definitely appreciate your engagement with our model. Let's see if this step helps clarify the situation! 🚀

**Comment by CedricLienhard** [comment]
Hi @glenn-jocher, thanks for your feedback!

I used the `--save-conf` and `--save-txt` flags but from what I see this only gives additional information about the **predictions** and not about the true labels. The model's predictions are actually okay and don't show these "skewed" rectangles (or parallelograms). These are only visible on the ground truth labels being displayed on their respective **VALIDATION** image. Again, these also don't appear on the training images, which is quite weird since one could expect that both the training and validation images would go through the same visualization process, but it appears that it isn't the case.

By looking a bit in the yolov8 source code, it seems that the validation samples are plotted using the `plot_val_samples()` method of validator.py, which uses `plot_image()` from plotting.py. I haven't had the time to deep dive into the code to exactly see where the issue is, but there might be an issue in how oriented bounding boxes are plotted.

**Comment by glenn-jocher** [comment]
Hi @CedricLienhard! 🌟

Thank you for diving deeper and providing these insights, and your effort in exploring the source code is genuinely appreciated!

You are correct; the `--save-conf` and `--save-txt` flags enrich the information regarding the predictions, not the true labels. Based on your observations, it indeed seems like the discrepancy arises from the visualization of ground truth labels for validation images. Since the predictions are accurate and properly displayed, the issue likely resides within the method used for visualizing the ground truth labels on validation data.

Given your findings, I agree that the differential handling of oriented bounding boxes in `plot_image()` during validation might be at the core of this. Though without a thorough investigation into `plot_val_samples()` and `plot_image()`, it's hard to pinpoint the precise cause.

I recommend raising an issue on the YOLOv8 GitHub repository detailing your observations. Including snippets from your investigation can help significantly. In the meantime, if the distortion in visualization doesn't affect evaluation metrics or prediction accuracy, it might be more of a visual anomaly than a functional issue with the model's performance or validation process.

Thanks again for reaching out and for your diligent work! Looking forward to seeing this get resolved. Keep up the great exploration! 🚀

**Comment by glenn-jocher** [comment]
@Laughing-q seems to be an issue with OBB val label plots not being true rectangles.

**Comment by glenn-jocher** [comment]
@CedricLienhard we've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub with `git pull` or `pip install -U ultralytics` to ensure your problem has not already been solved in the latest code version.
- ✅ **Unmodified** – Your problem must be reproducible using official Ultrlaytics code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/ultralytics/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by Laughing-q** [comment]
@CedricLienhard No worries, the val labels are all good, this is a known issue for val plotting only when images are rectangle.
**All the labels of val set are correct and validation process works properly without any issues.** :)

**Comment by glenn-jocher** [comment]
@Laughing-q is there a way to fix the known issue? Otherwise more users will probably raise similar bug reports.

**Comment by Laughing-q** [comment]
@glenn-jocher To visualize the labels on images, we'll need to convert obb from `xywhr` to `xyxyxyxy` format. And the distorted visualizations are caused by the minor error when operating `ops.xywhr2xyxyxyxy` on **normalized labels**， then the error scaled to imgsz-wise while denormalizing the coordinates.
https://github.com/ultralytics/ultralytics/blob/f2ddc963342e944214dc9d459a1e1dc4bf599aa1/ultralytics/utils/plotting.py#L842-L850
It's introduced by this PR: https://github.com/ultralytics/ultralytics/pull/7592 which fixed the incorrect visualized labels when image is rectangle.
Unfortunately it can't be fixed, because to fix the incorrect visualized labels in #7592 it has to operate `ops.xywhr2xyxyxyxy` on **normalized labels**.

**Comment by CedricLienhard** [comment]
@Laughing-q  I see, thanks for pointing the exact source of the issue and also for clarifying that the validation process works as intended, which was my main concern. 

This might be a stupid question but why not simply re-use the original validation labels when plotting, instead of propagating the errors coming from the normalization-denormalization process ? 

Cheers

**Comment by glenn-jocher** [comment]
@Laughing-q ah got it, ok thank you!

**Comment by Laughing-q** [comment]
@glenn-jocher @CedricLienhard Guys, good news! After a revisit of the obb dataset class and plotting code I eventually manged to fix the issue in #9899. :)
Also checkout the summary here: https://github.com/ultralytics/ultralytics/pull/9899#issue-2232581633
